Move memory deallocation out of critical section of ThreadGroupStatus::mutex

ZhiguoZh opened the pull request:
This commit optimizes the lock contention for ThreadGroupStatus::mutex by further shrinking the critical section for releasing the memory of ThreadGroupStatus:: finished_threads_counters_memory.

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- Performance Improvement

The performance experiments of SSB (Star Schema Benchmark) on the ICX device (Intel Xeon Platinum 8380 CPU, 80 cores, 160 threads) shows that this change could effectively decrease the lock contention for ThreadGroupStatus::mutex by **2%**, and increase the CPU utilization.

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
> Original changelog
In the original implementation, the memory of ThreadGroupStatus:: finished_threads_counters_memory is released by moving it to a temporary std::vector, which soon expired and gets destructed. This method is viable, however not straightforward enough. To enhance the code readability, this commit releases the memory in the vector by firstly resizing it to 0 and then shrinking the capacity accordingly.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

serxa commented on src/Interpreters/ThreadStatusExt.cpp:
```diff
@@ -337,7 +337,8 @@ void ThreadStatus::detachQuery(bool exit_if_already_detached, bool thread_exits)
     {
         std::lock_guard guard(thread_group->mutex);
-        auto stats = std::move(thread_group->finished_threads_counters_memory);
+        thread_group->finished_threads_counters_memory.resize(0);
+        thread_group->finished_threads_counters_memory.shrink_to_fit();
```
I'm not sure if these two lines are actually more readable than that one, but equivalent. But I feel like an idea of a code author was to avoid destruction under lock (and the current code is not doing this due to destruction order). Maybe you could improve this point here instead. First, move under lock, then destruct this heavy vector w/o locking the whole thread_group?

ZhiguoZh replied:
Thanks for sharing your insight! I will try to revise the patch accordingly.

ZhiguoZh replied:
Sergei, I've tried to revise the patch and evaluate the performance impact. The local performance test shows that this change could bring performance gain to some workloads, however, leads to slight regression in others. I'm working on the performance analysis and will update my findings once ready.

Thanks again!

ZhiguoZh committed:
Deallocate memory of profile events out of critical section

To further shrink the critical section for releasing memory of the
profile events (ProfileEventsCountersAndMemory), this commit puts
the dealloaction out of the critical section while keeping the
memory move under lock. This change could mitigate the contention
for ThreadGroupStatus::mutex.

ZhiguoZh commented:
Hi Sergei, as you suggested, moving the memory deallocation out of the critical section could further reduce the lock contention for ThreadGroupStatus::mutex, and when combined with another patch for relieving lock contention for ThreadPoolImpl::mutex, this change would deliver an overall performance improvement of 2.4%. As these 2 changes are not very closely related, I'd create another PR for it.

Could you help me review this patch? Thanks!

ZhiguoZh commented:
Hi Sergei, I've just triggered the tests on the latest base branch, and found the failed cases were due to the network issues (Bad gateway, etc.). Should we trigger another round of tests?

serxa commented:
Only a couple timeouts in tests
